### PR TITLE
[Algorithm] Discrete IQL

### PR DESCRIPTION
**Pull request by BY571**
## Description

Adding discrete IQL Loss + example script. 

Describe your changes in detail.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [X] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [X] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [X] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by BY571** [comment]
Performance tested on CartPole:

image

File: examples/iql/discrete_iql.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    frames_per_batch = cfg.collector.frames_per_batch
+    eval_rollout_steps = cfg.collector.max_frames_per_traj
+    sampling_start = start_time = time.time()
+    for tensordict in collector:
+        sampling_time = time.time() - sampling_start
+        pbar.update(tensordict.numel())
+        # update weights of the inference policy
+        collector.update_policy_weights_()
+
+        tensordict = tensordict.view(-1)
```
Comment (vmoens) [inline examples/iql/discrete_iql.py:103]: Safer to use reshape here

File: test/test_cost.py [hunk 2]
```diff
@@ -8950,6 +8951,768 @@ def test_iql_notensordict(
...
+            device=device,
+        )
+        return td
+
+    @pytest.mark.parametrize("num_qvalue", [1, 2, 4, 8])
+    @pytest.mark.parametrize("device", get_default_devices())
+    @pytest.mark.parametrize("temperature", [0.0, 0.1, 1.0, 10.0])
+    @pytest.mark.parametrize("expectile", [0.1, 0.5])
+    @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
+    def test_iql(
```
Comment (vmoens) [inline test/test_cost.py:9178]: We can rename this test_discrete_iql (as well as other tests)

File: torchrl/objectives/iql.py [hunk 3]
```diff
@@ -492,3 +493,352 @@ def make_value_estimator(self, value_type: ValueEstimators = None, **hyperparams
             "terminated": self.tensor_keys.terminated,
         }
         self._value_estimator.set_keys(**tensor_keys)
+
+
+class DiscreteIQLLoss(IQLLoss):
```
Comment (vmoens) [inline torchrl/objectives/iql.py:498]: Don't forget to put this in the doc!

**Review by vmoens — COMMENTED** [review]
LGTM just some very minor changes to make! Thanks a million

File: docs/source/reference/objectives.rst [hunk 4]
```diff
@@ -127,7 +127,7 @@ IQL
     :template: rl_template_noinherit.rst
 
     IQLLoss
-
+    DiscreteIQLLoss
```
Comment (vmoens) [inline docs/source/reference/objectives.rst:130]:
```suggestion
    DiscreteIQLLoss

```

**Comment by vmoens** [comment]
@BY571 we have a device exception that is being triggered

https://github.com/pytorch/rl/actions/runs/7530079022/job/20495721891?pr=1793

Do you know why?

**Review by vmoens — COMMENTED** [review]
LGTM thanks for this

**Review by vmoens — APPROVED** [review]

**Pull request merged**